src/main/java/com/direwolf20/buildinggadgets/common/building/Region.java

```diff
@@ -0,0 +1,240 @@
+        this.maxX += x;
+        this.maxY += y;
+        this.maxZ += z;
+        return this;
+    }
+
+    /**
+     * Expand the current region by the given value in all directions. Equivalent to {@link
+     * #grow(int, int, int)} with the given value for all 3 params. Negative values will shrink the region.
+     * <br/>
+     * Side lengths will be increased by 2 times the value of the parameter, since both min and max are changed.
+     * <br/>
+     * If contracting and the amount to contract by is larger than the length of a side, then the side will wrap (still
+     * creating a valid region - see samples on {@link #grow(int, int, int)}).
+     *
+     * @return this
+     */
+    public Region grow(int size) {
+        this.grow(size, size, size);
+        return this;
+    }
+
+    /**
+     * See {@link #grow(int, int, int)} subtracting instead of adding.
+     *
+     * @return this
+     */
+    public Region shrink(int x, int y, int z) {
+        return this.grow(-x, -y, -z);
+    }
+
+    /**
+     * See {@link #grow(int)} subtracting instead of adding.
+     *
+     * @return this
+     */
+    public Region shrink(int size) {
+        return this.grow(-size);
+    }
+
+    /**
+     * Create a new region with the intersecting part between the two regions.
+     *
+     * @return a new region
+     */
+    public Region intersect(Region other) {
+        int minX = Math.max(this.minX, other.minX);
+        int minY = Math.max(this.minY, other.minY);
+        int minZ = Math.max(this.minZ, other.minZ);
+        int maxX = Math.min(this.maxX, other.maxX);
+        int maxY = Math.min(this.maxY, other.maxY);
+        int maxZ = Math.min(this.maxZ, other.maxZ);
+        return new Region(minX, minY, minZ, maxX, maxY, maxZ);
+    }
+
+    /**
+     * Create a new region that encloses both regions that has the minimum volume.
+     *
+     * @return a new region
+     */
+    public Region union(Region other) {
+        int minX = Math.min(this.minX, other.minX);
+        int minY = Math.min(this.minY, other.minY);
+        int minZ = Math.min(this.minZ, other.minZ);
+        int maxX = Math.max(this.maxX, other.maxX);
+        int maxY = Math.max(this.maxY, other.maxY);
+        int maxZ = Math.max(this.maxZ, other.maxZ);
+        return new Region(minX, minY, minZ, maxX, maxY, maxZ);
+    }
+
+    /**
+     * @return this
+     */
+    @Override
+    public Region getBoundingBox() {
+        return this;
+    }
+
+    /**
+     * <p>
+     * The first result will have the minimum x, y, and z value. In the process it will advance first in positive z, and then x, and then y direction.
+     * In other words it will cover the first row with minimum x and y value, and then sweep through the lowest plane and work its way up.
+     * </p>
+     *
+     * @return {@link AbstractIterator}
+     * @implSpec starts at (minX, minY, minZ), ends at (maxX, maxY, maxZ)
+     */
+    @Override
+    public Iterator<BlockPos> iterator() {
```

Comment:
Could you make this implement Iterable? After that, could you expose a convenience Method: #stream() to make this usable as a Stream? (Use [StreamSupport#stream(spliterator,boolean)](https://docs.oracle.com/javase/8/docs/api/java/util/stream/StreamSupport.html).)

Streams are way more flexible, would remove the need for your composed iterator and the "AbstractTargetIterator" and are in general the Java 8 way to go...
You should replace any usages of your iterator with the Stream then...

Reply:
`IPlacementTarget` implements Iterable, so this is a Iterable too.